dsl/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+extern crate quote;
+extern crate litrs;
+extern crate proc_macro2;
+
+use quote::quote;
+
+use proc_macro2::TokenTree;
+use proc_macro2::TokenStream;
+use proc_macro2::TokenTree::Punct;
+use proc_macro2::TokenTree::Ident;
+use proc_macro2::token_stream::IntoIter;
+use std::iter::Peekable;
+
+use std::convert::TryFrom;
+use std::str::FromStr;
+use litrs::Literal;
+
+#[derive(PartialEq)]
+enum ExprType {
+    Nil, Int, Bool
+}
+
+// within the compiler, these generate tokens for
+// Value
+struct ParsedExpr {
+    t: ExprType,
+    tokens: TokenStream,
+}
+
+fn parse_literal(it: &mut Peekable<IntoIter>) -> ParsedExpr {
+    let token = it.next().expect("failed to parse literal, hit end of string");
+
+    match token {
+        Punct(ref x) => if x.as_char() == '(' {
+            let v = parse_expr(it);
+
+            it.next().expect("expected closing paren, hit end of string");
+            return v
+        },
+
+        // boolean literals
+        Ident(ref x) => match x.to_string().as_str() {
+            "true"  => return ParsedExpr{ t: ExprType::Bool, tokens: quote!{ true }},
+            "false" => return ParsedExpr{ t: ExprType::Bool, tokens: quote!{ false }},
+
+            // unary not
+            "not"   => {
+                let v = parse_literal(it).tokens;
+                return ParsedExpr{ t: ExprType::Bool, tokens: quote!{ Not::new(#v) }}
+            }
+            _ => panic!("error: unknown identifier"),
+        },
+
+        // kinda ugly but we fallthrough basically
+        _ => {},
+    }
+
+    // integer literals
+    if let Ok(Literal::Integer(i)) = Literal::try_from(token) {
+        let v = i.value::<u64>().expect("i don't think this integer fits?");
+        return ParsedExpr{ t: ExprType::Int, tokens: quote!{ #v }}
+    }
+
+    ParsedExpr{ t: ExprType::Nil, tokens: TokenStream::new() }
+}
+
+fn precedence(it: &mut Peekable<IntoIter>) -> Option<(i32, ExprType, &'static str)> {
```

Comment:
A minor gripe, but I feel such return signatures call for a custom enum instead to make it more readable.

Reply:
changed it to an `Option<Operator>` to avoid the uglier tuple, I'm curious if that's fine